src/common/mem.c

```diff
@@ -1573,4 +1573,428 @@ umem_tx_publish(struct umem_instance *umm, struct umem_rsrvd_act *rsrvd_act)
+	struct umem_checkpoint_data *chkpt_data_all;
+	struct umem_checkpoint_data *chkpt_data;
+	uint64_t                     committed_tx = 0;
+	uint64_t                     chkpt_id     = *out_id;
+	d_list_t                     free_list;
+	int                          i;
+	int                          rc;
+	int                          inflight = 0;
+
+	if (cache == NULL)
+		return 0; /* TODO: When SMD is supported outside VOS, this will be an error */
+
+	if (d_list_empty(&cache->ca_pgs_dirty))
+		return 0;
+
+	D_ASSERT(store != NULL);
+
+	D_INIT_LIST_HEAD(&free_list);
+	D_ALLOC_ARRAY(chkpt_data_all, MAX_INFLIGHT);
+	if (chkpt_data_all == NULL)
+		return -DER_NOMEM;
+
+	/** Setup the inflight IODs */
+	for (i = 0; i < MAX_INFLIGHT; i++) {
+		chkpt_data = &chkpt_data_all[i];
+		d_list_add_tail(&chkpt_data->cd_link, &free_list);
+		chkpt_data->cd_store_iod.io_regions = &chkpt_data->cd_regions[0];
+		chkpt_data->cd_sg_list.sg_iovs      = &chkpt_data->cd_iovs[0];
+	}
+
+	d_list_splice_init(&cache->ca_pgs_dirty, &cache->ca_pgs_copying);
+
+	/** First mark all pages in the new list so they won't be moved by an I/O thread.  This
+	 *  will enable us to continue the algorithm in relative isolation from I/O threads.
+	 */
+	d_list_for_each_entry(page, &cache->ca_pgs_copying, pg_link) {
+		/** Mark all pages in copying list first.  Marking them as waiting will prevent
+		 *  them from being moved to another list by an I/O operation.
+		 */
+		page->pg_waiting = 1;
+		if (store->stor_ops->so_wal_id_cmp(store, page->pg_last_inflight, chkpt_id) > 0)
+			chkpt_id = page->pg_last_inflight;
+	}
+
+	do {
+		/** first try to add up to MAX_INFLIGHT pages to the waiting queue */
+		while (inflight < MAX_INFLIGHT &&
+		       (page = d_list_pop_entry(&cache->ca_pgs_copying, struct umem_page,
+						pg_link)) != NULL) {
+			chkpt_data =
+			    d_list_pop_entry(&free_list, struct umem_checkpoint_data, cd_link);
+			D_ASSERT(chkpt_data != NULL);
+			page2chkpt(page, chkpt_data);
+
+			/** Presently, the flush API can yield.  Yielding is fine but ideally,
+			 *  we would like it to fail in such cases so we can run page2chkpt again.
+			 *  As it stands, we must set the copying bit here to avoid changes to the
+			 *  page.
+			 */
+			page->pg_copying = 1;
+			rc = store->stor_ops->so_flush_prep(store, &chkpt_data->cd_store_iod,
+							    &chkpt_data->cd_fh);
+			if (rc != 0) {
+				/** Just put the page back and break the loop */
+				page->pg_copying = 0;
+				d_list_add(&page->pg_link, &cache->ca_pgs_copying);
+				d_list_add(&chkpt_data->cd_link, &free_list);
+				break;
+			}
+
+			page->pg_last_checkpoint = page->pg_last_inflight;
+
+			rc = store->stor_ops->so_flush_copy(chkpt_data->cd_fh,
+							    &chkpt_data->cd_sg_list);
+			/** If this fails, it means invalid argument, so assertion here is fine */
+			D_ASSERT(rc == 0);
+
+			page->pg_copying = 0;
+
+			page_insert_sorted(store, page, &cache->ca_pgs_waiting);
```

Comment:
The idea is to keep the pipeline full.  Suppose MAX_INFLIGHT were 2 and you had two pages in the queue.  We will issue a wait for the checkpoint to commit.   By picking the lowest value to wait for, we can add another entry to the queue and kick it off while the higher checkpoint is in process of committing to the WAL.  Since the list is small, the overhead is pretty negligible.